Added variable which stores the recently used targets

cslux opened the pull request:
Hey Oleh Krehel,

I added a variable (`helm-make-recent-target-list`) which holds a list of the recently used targets. 
This variable is used for the various completion backends (helm/ivy/ido)

To get the next target stored in the list one has to type:
* **Helm** -> M-n
* **Ivy**     -> M-p
* **Ido**    -> M-p

Let me know what you think.

Regards,
Christian

PS: I will create another pull request, as I have changed also the functionality of the function
`helm-make-projectile`. 
After the changes, it will look first in the directory `helm-make-build-dir` for a makefile, when non-nil. Followed by `projectile-project-root` and `projectile-project-root/build`. 

Why should we look by default in `projectile-project-root/build`? Well, because it is common to build 
CMake/Autotools, (and probably some more), based project in a different directory, usually `build`.

cslux commented:
Oh, I did not know that github will append my other commit f97db41 to it.
Anyway, commit f97db41 changes the functionality of `helm-make-projectile` as stated in my previous comment.

Regards,
Christian

abo-abo commented:
I've edited your change heavily. A few points:

- I like to avoid the "merge ..." commits - they are useless. Just use `git rebase` to make nice modular (i.e. any single commit doesn't break the code) commits.

- I've renamed the var to `helm-make-target-history`, since that's the Emacs convention.

- I've removed the `:initial-input` stuff, since I think it's inappropriate for completion: with initial input you can't see all targets. Instead there's the `:preselect` method that works much better; both `helm` and `ivy` have it, only `ido` has to suffer a bit.

Anyway, thanks a lot for the change, I hope I didn't offend to much by amending it and to get more from you in the future:)

cslux commented:
> I like to avoid the "merge ..." commits - they are useless. Just use git rebase to make nice modular (i.e. any single commit doesn't break the code) commits.

Good point. I also was quite unhappy as I saw the pull-request (Yuck).

> I've renamed the var to helm-make-target-history, since that's the Emacs convention.

Oh, that was the first name I gave the variable, but changed it later :-/. I'll follow the emacs convention the next time. Thanks for the point.

> I've removed the :initial-input stuff, since I think it's inappropriate for completion: with initial input you can't see all targets. Instead there's the :preselect method that works much better; both helm and ivy have it, only ido has to suffer a bit.

I'm fine with that.

> Anyway, thanks a lot for the change, I hope I didn't offend to much by amending it and to get more from you in the future:)

I have absolutely no problem with that, in fact I like it when someone gives feedback. I'll try to get the most out of or feedback, thanks. You will get more from me, that's for sure :-)

cslux commented:
I personally would delete the duplicates from `helm-make-target-history`, before or after calling the completion backends. (see https://github.com/cslux/helm-make/commit/381a9ef540df8180fe0ea8d8418eb5ab36f8bc5a).

abo-abo commented:
> delete the duplicates from helm-make-target-history

Are there duplicates? Ivy deletes duplicates by default. Doesn't Helm do this as well?

cslux commented:
> Are there duplicates? Ivy deletes duplicates by default. Doesn't Helm do this as well?

Yes there are duplicates:

When `helm-make-completion-method` is `helm`

``` emacs-lisp
C-h v helm-make-target-history
("all" "clean" "all" "clean" "insta" "all")
```

After switching to `ivy`, and executing some more targets the variable
`helm-make-target-history` holds the following values.

``` emacs-lisp
(#("all" 0 3
   (ivy-index 8))
 #("clean" 0 5
   (ivy-index 0))
 #("install" 0 7
   (ivy-index 2))
 "insta")
```

Switching again back to `helm`, and executing some more targets the variable
`helm-make-target-history` holds the following values.

``` emacs-lisp
("clean" "all" "uninstall" "install" "clean"
 #("all" 0 3
   (ivy-index 8))
 #("clean" 0 5
   (ivy-index 0))
 #("install" 0 7
   (ivy-index 2))
 "insta")
```

That means, as you said, that `ivy` does remove the duplicates, but `helm` doesn't.
